README.md
## rtf R Package

An R package to produce RTF documents.

### Installation
From CRAN:

```
install.packages("rtf")
```

or from github:
```
library(devtools)
install_github("schaffman5/rtf")
```


### Usage
See the vignette for usage details: https://cran.r-project.org/web/packages/rtf/index.html

```
vignette("rtf")
```


NEWS
CHANGES IN 0.4-14 (2019-05-26)
-----------------------------
Updated with R.oo suggestions from Henrik Bengtsson. Added README.md and 
converted to RStudio project.

CHANGES IN 0.4-13 (2018-05-16)
-----------------------------
Added explicit imports and exports to resolve CRAN checks.

CHANGES IN 0.4-12 (2014-08-07)
-----------------------------
Added color table.

CHANGES IN 0.4-11 (2013-11-12)
-----------------------------
Changed to author to use Authors@R

CHANGES IN 0.4-10 (2013-08-20)
-----------------------------
Changed imports to include R.methodsS3.
Changed license to GPL.
Moved gsubfn to Suggests instead of Depends.
Removed dependency that forced X-windows start when the library was loaded.
Removed 'methods' dependency (thanks to Henrik Bengtsson for help).

CHANGES IN 0.4-9 (2013-08-18)
-----------------------------
Unreleased version.

CHANGES IN 0.4-8 (2013-06-04)
-----------------------------
Added support for etables.

CHANGES IN 0.4-7 (2012-08-22)
-----------------------------
Added table of contents support.
Added page number support in footers.
Added right justify for numeric values in tables.
Added full Greek alphabet to .convert function.

CHANGES IN 0.4-6 (2012-06-19)
-----------------------------
Added xtable to suggests list.

CHANGES IN 0.4-5 (2012-06-17)
-----------------------------
Corrected an incorrect Greek symbol substitution and added support for full Greek alphabet.
Fixed hidden method documentation with Rdoc annotation: @keyword -internal
Added package vignette.

CHANGES IN 0.4-4 (2012-04-11)
-----------------------------
Added documentation to output ggplot2 plots.
Added generic addPng method for inserting existing images.
Fixed output of some UTF-8 symbols method that were not being escaped properly.
Fixed a problem with the row.names parameter in addTable for data.frames.

CHANGES IN 0.4-3 (2011-10-03)
-----------------------------
Added NAMESPACE as required for packages in R-2.14.0.
Added parameter to rotate lattice plots in RTF output.
Added parameter for automatic column width detection in tables.
Fixed output of data.frame row names in addTable method.

CHANGES IN 0.4-2 (2011-09-05)
-----------------------------
Updates to documentation.  Added forest plot code.

CHANGES IN 0.4-1 (2011-09-04)
-----------------------------
Added license file and fixed an inappropriate is.nan call on a data frame.

CHANGES IN 0.4-0 (2011-09-04)
-----------------------------
Submission to CRAN.


tests/tests.R
library(rtf)
library(lattice)

output<-"test_rtf-package.doc"
png.res<-300

rtf<-RTF(output,width=8.5,height=11,font.size=10,omi=c(1,1,1,1))
addHeader(rtf,title="Test",subtitle="2011-08-15\n")
addPlot(rtf,plot.fun=plot,width=6,height=6,res=300, iris[,1],iris[,2])

# Try trellis plots

# single page trellis objects
addPageBreak(rtf, width=11,height=8.5,omi=c(0.5,0.5,0.5,0.5))

p <- histogram( ~ height | voice.part, data = singer, xlab="Height")
addTrellisObject(rtf,trellis.object=p,width=10,height=7.5,res=png.res)

p <- densityplot( ~ height | voice.part, data = singer, xlab = "Height")
addTrellisObject(rtf,trellis.object=p,width=10,height=7.5,res=png.res)

# multipage trellis object
p2<-xyplot(uptake ~ conc | Plant, CO2, layout = c(2,2))
addTrellisObject(rtf,trellis.object=p2,width=6,height=6,res=png.res)

addPageBreak(rtf, width=6.5,height=10,omi=c(0.5,0.5,0.5,0.5))
addParagraph(rtf,"\n\nHere's a new page with a custom size and margins.\n")

tab<-table(iris$Species,floor(iris$Sepal.Length))
names(dimnames(tab))<-c("Species","Sepal Length")
addParagraph(rtf,"\n\nTable based on a 'table' class object:\n")
addTable(rtf,tab,font.size=10,row.names=TRUE,NA.string="-",col.widths=c(1,rep(0.5,4)) )

addParagraph(rtf,"\n\nTable based on a 'data.frame' class with automatic column widths:\n")
tab<-as.data.frame(head(iris))
colnames(tab)<-gsub("\\."," ",colnames(tab))
addTable(rtf,tab,font.size=9,row.names=FALSE,NA.string="-")


tab<-head(as.data.frame(Seatbelts))
colnames(tab)<-c("car drivers killed","UK Driver Deaths","front-seat passengers killed or seriously injured","rear-seat passengers killed or seriously injured","distance driven (km)","petrol price","number of van drivers","Law in effect that month? (1/0)")
tab<-format(tab,digits=3)
addParagraph(rtf,"\n\nTable based on a 'data.frame' class with automatic column widths (exceeds page content width)\n")
addTable(rtf,tab,font.size=9,row.names=TRUE,NA.string="-")


fake.word<-function(all.upper=F,width.range=c(2,10)) {
	if(all.upper==FALSE) {
		return(paste(sample(c(LETTERS,letters),sample(width.range[1]:width.range[2])[1]),collapse=""))
	} else {
		return(paste(sample(c(LETTERS),sample(width.range[1]:width.range[2])[1]),collapse=""))
	}
}

fake.words<-function(word.max=3,all.upper=F,width.range=c(2,10)) {
	paste(replicate(sample(word.max)[1],fake.word(all.upper=all.upper,width.range=width.range)),collapse=" ")
}


# Single "word" data, first all uppercase, then mixed upper and lower case
ncols<-3
tab<-as.data.frame(matrix(replicate(ncols*5,fake.words(word.max=1,all.upper=T)),ncol=ncols))
colnames(tab)<-paste("C",c(1:ncols),sep="")
addParagraph(rtf,"\n\nAnother table based on a 'data.frame' class with automatic column widths (single-word/all upper case)\n")
addTable(rtf,tab,font.size=10,row.names=TRUE,NA.string="-")

ncols<-3
tab<-as.data.frame(matrix(replicate(ncols*5,fake.words(word.max=1,all.upper=F)),ncol=ncols))
colnames(tab)<-paste("C",c(1:ncols),sep="")
addParagraph(rtf,"\n\nAnother table based on a 'data.frame' class with automatic column widths (single-word/mixed case)\n")
addTable(rtf,tab,font.size=10,row.names=TRUE,NA.string="-")


# Multi-"word" data, first all uppercase, then mixed upper and lower case
ncols<-3
tab<-as.data.frame(matrix(replicate(ncols*5,fake.words(word.max=2,all.upper=T)),ncol=ncols))
colnames(tab)<-paste("C",c(1:ncols),sep="")
addParagraph(rtf,"\n\nAnother table based on a 'data.frame' class with automatic column widths (multi-word/all upper case)\n")
addTable(rtf,tab,font.size=10,row.names=TRUE,NA.string="-")

ncols<-4
tab<-as.data.frame(matrix(replicate(ncols*5,fake.words(word.max=3,all.upper=F)),ncol=ncols))
colnames(tab)<-paste("C",c(1:ncols),sep="")
addParagraph(rtf,"\n\nAnother table based on a 'data.frame' class with automatic column widths (multi-word/mixed case)\n")
addTable(rtf,tab,font.size=10,row.names=TRUE,NA.string="-")

# Try some UTF-8 output
addParagraph(rtf,"\n\nTry some UTF-8 output in a table\n")
addTable(rtf,data.frame(utf="\u2586"),font.size=10,row.names=FALSE,NA.string="-")

# Add a Greek substitutions
addParagraph(rtf,"\n\nTry some Greek letters\n&Alpha; &Beta; &Gamma; &Delta; &Epsilon; &Zeta; &Eta; &Theta; &Iota; &Kappa; &Lambda; &Mu; &Nu; &Xi; &Omicron; &Pi; &Rho; &Sigma; &Tau; &Upsilon; &Phi; &Chi; &Psi; &Omega; &alpha; &beta; &gamma; &delta; &epsilon; &zeta; &eta; &theta; &iota; &kappa; &lambda; &mu; &nu; &xi; &omicron; &pi; &rho; &sigmaf; &sigma; &tau; &upsilon; &phi; &chi; &psi; &omega;\n\n")


addNewLine(rtf)
addNewLine(rtf)
addSessionInfo(rtf)
done(rtf)

# Open in word processor
#system(paste("open",output))


# Test forest plot
tab<-data.frame(
	Label=c("Test1","Test2","Test3"),
	HR=c(1,2,0.45),
	Lower.CI=c(0.5,1.1,0.25),
	Upper.CI=c(2,3.5,0.9), 
	stringsAsFactors=FALSE, 
	check.names=FALSE)

# create forest plots by row
forest.plot.args<-list(xlim=c(0.1,5),width=3.0,height=0.3,cex=1,lwd=0.75,res=300)
tab$"HR Plot (log scale)"<-mapply(rtf.forest.plot,tab$HR,tab$Lower.CI,tab$Upper.CI,
			MoreArgs=forest.plot.args)

# rbind the x-scale to the table in the plot column
xscale<-rtf.forest.plot.xscale(xlim=c(0.1,5),width=3.0,height=0.3,cex=1,
			lwd=0.75,res=300)

tab<-data.frame(lapply(tab, as.character), 
			stringsAsFactors=FALSE, 
			check.names=FALSE)

tab<-rbind(tab,list("","","","",xscale))

# write the RTF output
rtf<-RTF("test.forest.plot.doc",width=8.5,height=11,font.size=10,omi=c(1,1,1,1))
addTable(rtf,tab,col.widths=c(0.75,0.75,0.75,0.75,3))
done(rtf)

# End of forest plot test









R/999.package.R
#########################################################################/**
# @RdocPackage rtf
#
# \encoding{latin1}
# 
# \description{
# A set of R functions to output Rich Text Format (RTF) files with high
# resolution tables and graphics that may be edited with a standard word
# processor such as Microsoft Word.  This is useful for compiling R
# results into a document for further editing or to be merged with another
# document.
#
# While Sweave and odfWeave provide nicely formatted outputs, the syntax
# can be difficult for beginners to master. Moreover, Sweave compilation
# to a PDF is not editable.  Other packages such as SWord and R2wd provide
# similar functionality but require the user to be running a Microsoft
# Windows OS.
# 
# This package is written in pure R and does not require leaving the R
# environment to write files. Data frames are converted to RTF tables and
# formatted as stored.  One important limitation of the RTF specification
# is that vector graphics output is limited to Windows Meta File (WMF) and
# Enhanced Meta File (EMF) formats.  Because these formats are not
# supported across platforms, this package currently only supports RTF
# embedding of PNG plots and images.  To ensure high quality reports, the
# resolution may be specified when writing the RTF output.
# }
#
# \section{Requirements}{
#   This package depends on the \pkg{R.oo} package.
# }
#
# @author
# 
# \references{
# [1] \url{http://en.wikipedia.org/wiki/Rich_Text_Format}\cr
#
# [2] \url{http://latex2rtf.sourceforge.net/rtfspec_7.html#rtfspec_paraforprop}\cr
# }
#
# \keyword{ rtf }
#
#*/#########################################################################


R/999.NonDocumentedObjects.R
###########################################################################/**
# @RdocDocumentation "Non-documented objects"
#
# @alias view
# @alias addTable
# @alias done
# @alias addPageBreak
# @alias startParagraph
# @alias endParagraph
# @alias addHeader
# @alias addParagraph
# @alias addText
# @alias addNewLine
# @alias decreaseIndent
# @alias increaseIndent
# @alias setFontSize
# @alias addPlot
# @alias addPng
# @alias addTrellisObject
# @alias addSessionInfo
#
# @alias rtf.forest.plot.xscale
#
#
# \description{
#   This page contains aliases for all "non-documented" objects that 
#   \code{R CMD check} detects in this package. 
#
#   Almost all of them are \emph{generic} functions that have a specific 
#   document for the corresponding method coupled to a specific class. 
#   Other functions are re-defined by \code{setMethodS3()} to 
#   \emph{default} methods. Neither of these two classes are non-documented
#   in reality.
# }
#
# @author
#
# @keyword internal
#*/###########################################################################

############################################################################
# HISTORY:
# 2011-09-04
# o Created to please R CMD check.
############################################################################


R/forest.plot.R
#########################################################################/**
# @RdocFunction rtf.forest.plot
#
# @title "Get an RTF encoded forest plot"
#
# \description{
#	@get "title".
# }
#
# @synopsis
#
# \arguments{
# 	\item{x}{x (e.g. hazard ratio).}
# 	\item{min}{Minimum whisker (e.g. lower bound of 95\% hazard ratio CI).}
# 	\item{max}{Maximum whisker (e.g. upper bound of 95\% hazard ratio CI).}
# 	\item{xlim}{A @vector specifying the x limits.}
# 	\item{width}{Plot width in inches.}
# 	\item{height}{Plot height in inches.}
# 	\item{cex}{A numerical value giving the amount by which plotting text and symbols should be magnified relative to the default.}
# 	\item{lwd}{Line width.}
# 	\item{res}{Output resolution in dots per inch.}
# }
#
# \details{
# 	Create a forest plot and convert PNG to RTF code.  This is useful for
# 	embedding into a data frame of hazard ratios and then writing an
# 	RTF output file.  See the example below for usage.
# }
#
# \examples{
# tab<-data.frame(
#	Label=c("Test1","Test2","Test3"),
# 	HR=c(1,2,0.45),
# 	Lower.CI=c(0.5,1.1,0.25),
# 	Upper.CI=c(2,3.5,0.9), 
# 	stringsAsFactors=FALSE, 
# 	check.names=FALSE)
#
# # create forest plots by row
# forest.plot.args<-list(xlim=c(0.1,5),width=3.0,height=0.3,cex=1,lwd=0.75,res=300)
# tab$"HR Plot (log scale)"<-mapply(rtf.forest.plot,tab$HR,tab$Lower.CI,tab$Upper.CI,
#			MoreArgs=forest.plot.args)
#
# # rbind the x-scale to the table in the plot column
# xscale<-rtf.forest.plot.xscale(xlim=c(0.1,5),width=3.0,height=0.3,cex=1,
# 			lwd=0.75,res=300)
#
# tab<-data.frame(lapply(tab, as.character), 
# 			stringsAsFactors=FALSE, 
#			check.names=FALSE)
#
# tab<-rbind(tab,list("","","","",xscale))
#
# # write the RTF output
# rtf<-RTF("test_rtf.forest.plot.doc",width=8.5,height=11,font.size=10,omi=c(1,1,1,1))
# addTable(rtf,tab,col.widths=c(0.75,0.75,0.75,0.75,3))
# done(rtf)
# }
#
# @author
#*/#########################################################################
rtf.forest.plot<-function(x=1.25,min=0.5,max=2,xlim=c(0.1,12),
	width=3.0,height=0.3,cex=1,lwd=0.75,res=300) {
	
	tmp.file<-tempfile("temp_forest_plot")
	ret<-.rtf.plot(.forest.plot,tmp.file=tmp.file,width=width,height=height,res=res,x=x,min=min,max=max,xlim=xlim,cex=cex,lwd=lwd)
	if(file.exists(tmp.file) ) {
		unlink(tmp.file)
	}
	ret
}

rtf.forest.plot.xscale<-function(xlim=c(0.1,12),
	width=3.0,height=0.3,cex=1,lwd=0.75,res=300) {
	tmp.file<-tempfile("temp_forest_xscale")
	ret<-.rtf.plot(.forest.plot.scale,tmp.file=tmp.file,width=width,height=height,res=res,xlim=xlim,cex=cex,lwd=lwd)
	if(file.exists(tmp.file) ) {
		unlink(tmp.file)
	}
	ret
}

######################################################################################

.forest.plot<-function(x=1.25,min=0.5,max=2,xlim=c(0.1,12),cex=1,lwd=0.75) {
	par(oma=c(0,0,0,0),mar=c(0,0,0,0))
	plot(c(1,1),type="n",axes=FALSE,xlab="",ylab="",xlim=xlim,ylim=c(1,1),log="x",xaxt="n",yaxt="n",bty="l")

	abline(v=1,lwd=0.5,col="black") # tick marks at '1' behind the plot glyphs

	if(!is.na(min)) { arrows(x,1,min,1,angle=90,lwd=lwd,length=0.05) }
	if(!is.na(max)) { arrows(x,1,max,1,angle=90,lwd=lwd,length=0.05) }	
	if(!is.na(x)) { points(x,1,pch=21,bg="black",cex=cex) }
}

.forest.plot.scale<-function(xlim=c(0.1,12),cex=1,lwd=0.75,res=300) {
	par(oma=c(0,0,0,0),mar=c(0,0,0,0))
	plot(c(1,1),type="n",axes=FALSE,xlab="",ylab="",xlim=xlim,ylim=c(0.95,1.05),log="x",xaxt="n",yaxt="n",bty="l")
	ticks<-axTicks(1)
	y.max=1.025
	sapply(ticks,function(x,y.max){lines(x=c(x,x),y=c(1,y.max),lwd=lwd)},y.max=y.max)
	lines(c(ticks[1],ticks[length(ticks)]),c(y.max,y.max),lwd=lwd)
	text(ticks,1,ticks,pos=1,offset=0.25,xpd=NA,cex=cex)
}
